fl4health/clients/fenda_client.py

```diff
@@ -1,12 +1,229 @@
+        as positive pairs while reducing the similarity between the current local features and current global
+        features as negative pairs.
+        """
+        assert isinstance(self.temperature, float)
+        assert len(local_features) == len(old_local_features)
+        posi = self.cos_sim(local_features, old_local_features)
+        logits = posi.reshape(-1, 1)
+        assert len(local_features) == len(global_features)
+        nega = self.cos_sim(local_features, global_features)
+        logits = torch.cat((logits, nega.reshape(-1, 1)), dim=1)
+        logits /= self.temperature
+        labels = torch.zeros(local_features.size(0)).to(self.device).long()
+
+        return self.ce_criterion(logits, labels)
+
+    def get_perFCL_loss(
+        self,
+        local_features: torch.Tensor,
+        old_local_features: torch.Tensor,
+        global_features: torch.Tensor,
+        old_global_features: torch.Tensor,
+        aggregated_global_features: torch.Tensor,
+    ) -> Tuple[torch.Tensor, torch.Tensor]:
+        """
+        PerFCL loss consists of two contrastive losses.
+        First one airms to enhance the similarity between the current global features and aggregated global feature
+        as positive pairs while reducing the similarity between the current global features and old global
+        features as negative pairs.
+        Second one airms to enhance the similarity between the current local features and old local feature
+        as positive pairs while reducing the similarity between the current local features and aggregated global
+        features as negative pairs.
+        """
+
+        assert isinstance(self.temperature, float)
+        assert len(global_features) == len(aggregated_global_features)
+        posi = self.cos_sim(global_features, aggregated_global_features)
+        logits_max = posi.reshape(-1, 1)
+        assert len(global_features) == len(old_global_features)
+        nega = self.cos_sim(global_features, old_global_features)
+        logits_max = torch.cat((logits_max, nega.reshape(-1, 1)), dim=1)
+        logits_max /= self.temperature
+        labels_max = torch.zeros(local_features.size(0)).to(self.device).long()
+
+        assert len(local_features) == len(old_local_features)
+        posi = self.cos_sim(local_features, old_local_features)
+        logits_min = posi.reshape(-1, 1)
+        assert len(local_features) == len(aggregated_global_features)
+        nega = self.cos_sim(local_features, aggregated_global_features)
+        logits_min = torch.cat((logits_min, nega.reshape(-1, 1)), dim=1)
+        logits_min /= self.temperature
+        labels_min = torch.zeros(local_features.size(0)).to(self.device).long()
+
+        return self.ce_criterion(logits_min, labels_min), self.ce_criterion(logits_max, labels_max)
+
+    def compute_loss(self, preds: Dict[str, torch.Tensor], target: torch.Tensor) -> Losses:
+        if self.old_global_module is None:
```

Comment:
We have a "catch all" fallback in the form of the else statement at the end of this function. It seems like if this module is defined and we still get down to the else statement below, something has been "weirdly" configured and we should at least warn the user about it? Alternatively, we could just eliminate this else and let the catch all happen if all the weight values are `None`